shards/modules/physics/debug.cpp

```diff
@@ -0,0 +1,305 @@
+#include <shards/modules/gfx/shards_types.hpp>
+#include <shards/modules/gfx/gfx.hpp>
+#include <shards/modules/gfx/gizmos/context.hpp>
+#include <shards/linalg_shim.hpp>
+#include <gfx/gizmos/shapes.hpp>
+#include <gfx/drawables/mesh_drawable.hpp>
+#include <gfx/mesh.hpp>
+#include <shards/core/shared.hpp>
+#include <shards/core/params.hpp>
+#include <shards/common_types.hpp>
+#include "core.hpp"
+#include <Jolt/Renderer/DebugRenderer.h>
+#include <cmath>
+
+namespace shards::Physics {
+
+struct DebugRenderer : public JPH::DebugRenderer {
+  struct GFXBatch : public JPH::RefTargetVirtual {
+    gfx::MeshPtr mesh;
+
+    virtual void AddRef() override { refCount.fetch_add(1, std::memory_order::relaxed); }
+    virtual void Release() override {
+      if (--refCount == 0)
+        delete this;
+    }
+
+  private:
+    std::atomic_uint32_t refCount = 0;
+  };
+
+  gfx::GizmoRenderer *gizmoRenderer;
+  gfx::ShapeRenderer *shapeRenderer;
+  gfx::WireframeRenderer *wireframeRenderer;
+  std::vector<gfx::DrawablePtr> drawables;
+
+  DebugRenderer() { Initialize(); }
+
+  void setGizmoRenderer(gfx::GizmoRenderer &gizmoRenderer) {
+    this->gizmoRenderer = &gizmoRenderer;
+    this->shapeRenderer = &gizmoRenderer.getShapeRenderer();
+  }
+
+  void setWireframeRenderer(gfx::WireframeRenderer &wireframeRenderer) { this->wireframeRenderer = &wireframeRenderer; }
+
+  void DrawLine(JPH::RVec3Arg inFrom, JPH::RVec3Arg inTo, JPH::ColorArg inColor) override {
+    shapeRenderer->addLine(toLinalg(inFrom), toLinalg(inTo), toLinalgLinearColor(inColor), 1.0f);
+  }
+  void DrawTriangle(JPH::RVec3Arg inV1, JPH::RVec3Arg inV2, JPH::RVec3Arg inV3, JPH::ColorArg inColor,
+                    ECastShadow inCastShadow = ECastShadow::Off) override {
+    shapeRenderer->addSolidTriangle(toLinalg(inV1), toLinalg(inV2), toLinalg(inV3), toLinalgLinearColor(inColor));
+  }
+  void DrawText3D(JPH::RVec3Arg inPosition, const std::string_view &inString, JPH::ColorArg inColor = JPH::Color::sWhite,
+                  float inHeight = 0.5f) override {
+    float autoSize = gizmoRenderer->getConstantScreenSize(toLinalg(inPosition), inHeight * 14.0f * 5.0f);
+    gizmoRenderer->addTextBillboard(toLinalg(inPosition), inString, toLinalgLinearColor(inColor), autoSize, true);
+  }
+
+  Batch CreateTriangleBatch(const Triangle *inTriangles, int inTriangleCount) override {
+    gfx::MeshPtr mesh = std::make_shared<gfx::Mesh>();
+    gfx::MeshFormat fmt;
+    fmt.primitiveType = gfx::PrimitiveType::TriangleList;
+    fmt.vertexAttributes = {
+        gfx::MeshVertexAttribute("position", 3, gfx::StorageType::Float32),
+        gfx::MeshVertexAttribute("normal", 3, gfx::StorageType::Float32),
+        gfx::MeshVertexAttribute("texCoord0", 2, gfx::StorageType::Float32),
+        gfx::MeshVertexAttribute("color", 4, gfx::StorageType::UNorm8),
+    };
+    mesh->update(fmt, inTriangles, sizeof(Vertex) * 3 * inTriangleCount, nullptr, 0);
+
+    GFXBatch *batch = new GFXBatch();
```

Comment:
// Consider using std::unique_ptr instead of raw pointer for GFXBatch
GFXBatch *batch = new GFXBatch();
// Could be:
std::unique_ptr<GFXBatch> batch = std::make_unique<GFXBatch>();